### PR TITLE
[CON-1330] Add support for document cookie for interop with phc-login

### PR DESCRIPTION
**Pull request by xdumaine**

### CODE REVIEW

**Review by lars-mortenson — DISMISSED** [review]

File: src/LOAuth.ts [hunk 1]
```diff
@@ -112,6 +112,42 @@ class LOAuth {
...
+        expires: storedData.expiresAt
+      });
+      this.token.expiresIn(new Date(storedData.expiresAt - Date.now()));
+      this._storeTokenData(this.token);
+
+      // delete the cookie after it's read
+      document.cookie = `${AUTH_STORAGE_KEY}=;domain=.${storedData.cookieDomain};Max-Age=-9999;path=/;secure`;
+    } catch (err) {
+      return;
+    }
```
Comment (smolster) [inline src/LOAuth.ts:148]: Should we log, or callback somehow in the error case?

**Review by smolster — APPROVED** [review]

File: src/LOAuth.ts (hunk 1 above)
Comment (xdumaine) [reply src/LOAuth.ts:148]: Good catch, I didn't want to log to console directly and this lib doesn't have a logger object option. I figured since it'll just fallback to the normal flow after, I could silently ignore.

File: test/LOAuth.test.ts [hunk 2]
```diff
@@ -338,6 +341,22 @@ describe('with auth successfully created', () => {
     expect(globals.window.fetch).toBeCalledTimes(0);
   });
 
+  test('refreshAccessToken uses the token from storage if it exists', async () => {
```
Comment (xdumaine) [inline test/LOAuth.test.ts:344]:
```suggestion
  test('refreshAccessToken uses the token from the domain document cookie if it exists', async () => {
```

**Comment by austin-rausch** [comment]
I think this might not actually be compatible with what we'd like to do, but thats okay we can figure something else out and you guys can go forward with this.

There's a couple things we want to accomplish:

- Have a long lived cookie (either the access token directly or something that can be exchanged for the access token) so we can server side render pages for the user
- Allow said cookie to be read by all lifeomic domains (assuming they're the same user pool & the token is valid), E.G. `marketplace.dev.lifeomic.com` <-> `apps.dev.lifeomic.com` <-> `wellness.dev.lifeomic.com` etc. etc.

I was actually thinking of setting an http only cookie [here](https://github.com/lifeomic/apps-auth/blob/master/lambda/routes/api.js#L318), either directly as the auth result or something that is secure & random enough to be exchanged for the auth result, (since we're limited to ~4096 bytes per cookie, and splitting the tokens into multiple cookies could get complicated).

**Comment by xdumaine** [comment]
@austin-rausch I think that makes sense. if it's intended to be longer lived than just setting tokens for a pwa in local storage, then http-only is definitely the way to go. @macnaughton mentioned something similar.

**Review by smolster — DISMISSED** [review]

**Comment by xdumaine** [comment]
I have this running in fed in the connect-web-console and it's working properly.

File: src/LOAuth.ts [hunk 3]
```diff
@@ -112,6 +112,42 @@ class LOAuth {
...
+   * applications to exchange auth tokens over different subdomains
+   * see phc-login/src/util/handleAuthResponse
+   */
+  private _getDomainCookieAuthState() {
+    try {
+      if (!document.cookie) {
+        return;
+      }
+      const cookie = document.cookie
+        .split('; ')
```
Comment (mdlavin) [inline src/LOAuth.ts:126]: Manual cookie parsing surprises me. Is there not a good cookie util in our apps?

File: src/LOAuth.ts [hunk 4]
```diff
@@ -112,6 +112,42 @@ class LOAuth {
...
+      }
+
+      const storedData = JSON.parse(value);
+      this.token = this.client.createToken({
+        access_token: storedData.accessToken,
+        refresh_token: storedData.refreshToken,
+        token_type: 'Bearer',
+        expires: storedData.expiresAt
+      });
+      this.token.expiresIn(new Date(storedData.expiresAt - Date.now()));
```
Comment (mdlavin) [inline src/LOAuth.ts:141]: Is this Date logic write?  I thought the `Date` constructor accepted an Epoch time, but it looks like it's being pass a duration in ms

File: src/LOAuth.ts [hunk 5]
```diff
@@ -112,6 +112,42 @@ class LOAuth {
...
+        .split('; ')
+        .find((cookie) => cookie.startsWith(`${AUTH_STORAGE_KEY}=`));
+
+      const value = cookie && cookie.split('=')[1];
+      if (!value) {
+        return;
+      }
+
+      const storedData = JSON.parse(value);
+      this.token = this.client.createToken({
```
Comment (mdlavin) [inline src/LOAuth.ts:135]: Not blocking, but I find the pattern of this method modifying `this.token` and then line 190 checking that something happened a little hard to follow.

If I was writing it, I would have returned the token and set it in `refreshAccessToken` so that all `this.token` setting lives together

File: src/LOAuth.ts (hunk 4 above)
Comment (xdumaine) [reply src/LOAuth.ts:141]: good catch.

File: src/LOAuth.ts (hunk 3 above)
Comment (xdumaine) [reply src/LOAuth.ts:126]: there's not that I could find, and this seemed simple enough, but maybe it's worth finding a lib?

File: src/LOAuth.ts (hunk 5 above)
Comment (xdumaine) [reply src/LOAuth.ts:135]: that's reasonable, I can make that change.

File: src/LOAuth.ts (hunk 3 above)
Comment (mdlavin) [reply src/LOAuth.ts:126]: It's on the edge. I'm not sure if adding a new lib is worth it

File: src/LOAuth.ts [hunk 6]
```diff
@@ -112,6 +112,42 @@ class LOAuth {
     return btoa(JSON.stringify(state));
   }
 
+  /**
+   * This method looks for a cookie set by PHC-Login that allows the
```
Comment (xdumaine) [inline src/LOAuth.ts:116]: I'm going to rework this to have this lib set the cookie as well, and use the cookie set method inside phc-login to avoid this disjointedness.

File: src/LOAuth.ts [hunk 7]
```diff
@@ -112,6 +112,67 @@ class LOAuth {
...
+        refresh_token: storedData.refresh_token,
+        token_type: 'Bearer'
+      });
+      token.expiresIn(storedData.expires - Date.now());
+
+      // the cookie clientId takes precedent over configured options
+      this.clientOptions.clientId = storedData.clientId;
+
+      // delete the cookie after it's read
+      document.cookie = `${AUTH_STORAGE_KEY}=;domain=.${storedData.cookieDomain};Max-Age=-9999;path=/;secure`;
```
Comment (aecorredor) [inline src/LOAuth.ts:164]: `Max-Age=-9999` is what does the delete trick here right? Might be worth commenting that

File: package.json [hunk 8]
```diff
@@ -33,7 +33,7 @@
     "tslint": "^5.12.0",
     "tslint-config-prettier": "^1.17.0",
     "tslint-config-semistandard": "^7.0.0",
-    "typescript": "^3.2.2"
+    "typescript": "^3.7.5"
```
Comment (mdlavin) [inline package.json:36]: Bonus 🔥

**Review by mdlavin — APPROVED** [review]

File: package.json (hunk 8 above)
Comment (xdumaine) [reply package.json:36]: Ha yeah, I noticed a weird type error and this fixed it.

File: src/LOAuth.ts (hunk 7 above)
Comment (xdumaine) [reply src/LOAuth.ts:164]: 👍🏽 will add in the follow up
